CONTRIBUTING.md: establish initial automation/AI/LLM policy (#514587)
NixOS/nixpkgs: CONTRIBUTING.md

Reviewer: (here, because it's a reply to Emily's header message, not this file's contents)

> we would not want to ban the use of accessibility tools that often use LLMs like machine translation, speech to text, text to speech, and OCR

accessibility tools don't generate code, which is what this policy should regulate, they only interpret existing media, such as reading the given text or type from the person's voice. a contribution from a visually impaired person, who reads with LLM-powered tools, but codes everything by hand, isn't LLM-generated. it is certainly possible to restrict LLM usage and not restrict accessibility tools

original replies:
- https://github.com/NixOS/nixpkgs/pull/514587#issuecomment-4341695676
- https://github.com/NixOS/nixpkgs/pull/514587#issuecomment-4343186398

Reviewer 2: A coding agent can itself also be an accessibility tool, if someone cannot type for any reason (e.g. RSI).

Reviewer: RSI meaning repetitive strain injury?

anyway, once again, see my points in linked messages: there's a difference between voice-typing word by word (and knowing what you do) and "hey, claude, make this thing for me"

Reviewer 3: I'd like to highlight https://github.com/NixOS/nixpkgs/issues/516544 as a recent AI related controversy likely needing a policy, but not covered by this current PR

Reviewer 4: The emoji reactions might not be a representative sample but it appears like people overwhelmingly support the name change.

Please don't blow up situations where progress is happening but is merely not going fast enough for your tastes.
Honking in a traffic jam will not make you go any faster.

> # Automation/AI policy

Reviewer 5: Hey all,

I fear that this policy is still too permissive. I see the effort put into guardrails to guide LLM users to use their tools responsibly, but am scared that this will still encourage a flood of low effort contributions.

As you say, we are fundamentally bottlenecked on manual review. I am not enthusiastic about reviewing code that could be AI generated. I do not want to spend my time explaining why something fails, and how it can be fixed, to someone who will just pass it along to their chatbot. I do not think that allowing LLM generated content, even clearly labled, will help the burden of manual reviewers.

I believe automated, deterministic, auditable tools fundamentally differ from LLMs. While I agree that holding them to at least the same standards is a good starting point, I believe they should be held to a much stricter one - if allowed at all.

That said, I appreciate that a stance is being taken. Thank you for your transparency and request for feedback.

Reviewer 5: I, and many other maintainers have had multiple workflow disrupting interactions with people using AI as a substitute for thinking. Many contributors will not review, advise, or otherwise knowingly interact with any contribution that has been generated by an LLM. That will not change, regardless of what policy is implemented.

I understand that this is not the case for everyone here, and have had many discussions with people on both sides of the fence. One of the arguments I hear the most is that banning AI all-together is unenforceable. Or that a full ban will alienate new contributors. These are not reasons to implement a more permissive policy. We want new contributors to write their own code, be corrected, and learn from the experience. We want our energy to benefit the community.

Enforcing an AI ban will not be impossible. As [mentioned here](https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3167271117), banning AI is doable. In most situations it is immediately obvious when someone has used an LLM. If they are dishonest about their use, this can easily be escalated and verified. Multiple other large, community governed projects have chosen for a stricter policy, if not an outright ban (see [Wikipedia](https://en.wikipedia.org/wiki/Wikipedia:Artificial_intelligence_resources#Policies), [Gentoo](https://wiki.gentoo.org/wiki/Project:Council/AI_policy)).

Something similar to the [proposed rustlang AI policy](https://github.com/rust-lang/rust-forge/pull/1040) is much more in line with our [community values](https://nixos.org/governance/#community-values-people). Our policy should aim to lighten the load of reviewers, and encourage human learning.

It is imperative that we implement a policy, and these discussions are prone to stall. I urge the core team to close the discussion after the two week mark.

Author: We absolutely agree that contributions and reviews being passed unthinkingly back and forth to LLMs is not acceptable, and this policy is very much intended to forbid that; I’ve said more about this and other matters you’ve raised at https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175207606, so I’ll point there to avoid duplicating responses :)

> # Automation/AI policy

Reviewer 6: I know I'll get push-back on this, but I really think it's worth including at least a minimal AGENTS.md in the repository that just highlights these standards, if only because then we can enlist the agents themselves to help inform users about the expectations and responsibilities they have to adhere to as contributors.

Reviewer 6: They will read CONTRIBUTING.md at times, and I have a global instruction on my system for my agent to do so, but AGENTS.md is directly injected into the context window as part of the system prompt (or at least automatically injected into the context window at the start of all conversations) so they don't even *have* to read it.

Reviewer 7: I'll note that figuring out this sort of thing was the point of [one of the GSoC proposals](https://github.com/NixOS/GSoC/pull/32) that didn't make the cut. It might be worth revisiting, or even suggesting a stub `AGENTS.md` that has the shape you're considering--a strawman would go a long way. :)

Reviewer 8: might adding such a file like this not defeat the spirit of the stated brown M&M test though?

> # Automation/AI policy

Reviewer 9: Overall feeling positive about the proposal, will start with some stylistic remarks

```suggestion
# Automation and "AI" policy
```

1. Forward-slashes like these generally look like lazy writing
2. "AI" is a marketing term, I suggest we keep it quoted, and later elaborate, once, that what we mean is something along the lines of "data-driven programs"

Author: While I don’t love the slashes, I think “and” here is misleading, as the policy considers automation in general quite uniformly and treats the latter a subset of the former. I think adding the quotes would be more likely to confuse people than help; in the actual body text, “LLM‐based AI tools” is used consistently, which I think should be sufficiently precise, but the title should aim for obviousness and discoverability, which means sticking to common terminology. (Otherwise, it could just be the “automation policy”.)

A non‐standard term like “data-driven programs” seems like it would invite more bikeshedding over the boundaries. I’m tempted to say this and https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3163362758 cancel each other out 😅

Reviewer 9: >  I think “and” here is misleading

"And" is not always disjoint?

> “LLM‐based AI tools” is used consistently

While at it, I'd suggest: `"AI" and (L)LM-based tools`. This way we highlight that the policy isn't about any particular architecture or implementation, but about certain properties (complexity, poor inspectability, poor predictability, &c) of these tools.

> A non‐standard term like “data-driven programs”

It is somewhat common[^1][^2][^3], in any case less non-standard than "AI" the way it's used today. `"AI"` is good in a title.

[^1]: https://en.wikipedia.org/wiki/Data-driven_model
[^2]: https://www.youtube.com/watch?v=a13aqr07tJ4
[^3]: https://archive.org/details/Redwood_Center_2014_09_24_Alyosha_Efros

> # Automation/AI policy

Reviewer 10: Hi folks,

TL;DR: We probably shouldn't allow LLM-usage for the vast majority of contributions

I think something that folks in this issue are missing is the social component of open source software development. I'm assuming the vast majority of contributors to Nixpkgs are hobbyists or volunteers who are looking to give-back to the community/open source or learn more about Nix and its workings. While there are probably some folks who are paid to work on Nixpkgs in some capacity, I imagine it is a very, very, very small fraction of the total contributor base. Essentially, NixOS/Nixpkgs is not a production codebase and there are not economic incentives for most people to work on it efficiently.

By the way, this comment applies to code generation tools and not accessibility tools, which I fully support for obvious reasons.

But there's a social contract that we all accept when making contributions to a project or reviewing contributions: code in exchange for knowledge. If I make the (honest) effort to write code and contribute it to a project, I deserve to be corrected and taught if I make a mistake (through reviews, CI, documentation, etc.). Likewise, reviewers deserve an honest attempt to address the feedback and improve their contributions (present and future). By other people investing time in reviews to my code, they are investing in me and my future development and contribution to Nixpkgs. When I invest my (rare) time into someone's review, I am investing in them being future contributors and in their learning. This is why open source works.

LLMs short-circuit this social contract. LLMs do not learn. When someone submits a vibe-coded contribution to initialize a package, they are probably just going to feed your review straight back into claude code/codex/cursor/whatever and push the PR without any second thought (especially for one-off/drive-by contributions). The human will have learned and accomplished nothing except burning tokens. I personally will not tolerate this and will refuse to engage with contributors who are known to do this.

At the end of the day, contributing to Nixpkgs is not hard. It does not take a genius to figure out how to copy a template and change some lines. But there is real learning that occurs for a human that is reviewing and responding to feedback (be it created by a human or by deterministic CI). If you are unwilling to learn to do this yourself and learn, do we as a community really want you as a contributor or your code? I don't think so. You'll just make more work for me and other people who do understand and care later on.

Advocates of allowing LLM generated contributions may cite increased productivity and spending less time on contributions as a reason to allow these tools. I disagree. The vast majority of contributors are not being paid to work on Nixpkgs. There's not an economic (only time) incentive for us to work faster. We don't report to any shareholders or managers. If someone wants to spend less time on Nixpkgs, then they should do less, not do more lower-quality work. We have enough work to sort through as it is. 

That being said, I do think LLMs can be used properly for experienced contributors making certain kinds of changes. Like mentioned in another thread, tedious tree-wide changes can be easily made with LLMs in a way that deterministic tools struggle with. I think this is a perfectly practical use case (although deterministic solutions are preferable for environmental, ethical, and legal reasons). I think properly disclosing usage is important and appropriate here, too.

I am not staunchly anti LLM. I will be working for an AI company this summer. I use AI coding tools for my job (I do not use them for open source, school, or pet projects). For better or for worse, "AI" coding is probably here to stay (whether through local models, frontier labs, or otherwise). Programming and open source more broadly will change. It is what it is. But LLMs are not the solution to every technical problem. There is real value is learning, building experience, and the social aspect of open source. Let's not lose that for some perceived productivity gains and to gain lazy code that will ultimately become someone else's responsibility.

Of course, there are also a litany of [environmental](https://spectrum.ieee.org/state-of-ai-index-2026), [copyright](https://www.quippd.com/writing/2026/04/08/ai-code-is-hollowing-out-open-source-and-maintainers-are-looking-the-other-way.html), [ethical](https://hai.stanford.edu/news/covert-racism-ai-how-language-models-are-reinforcing-outdated-stereotypes), [legal](https://www.nytimes.com/2025/09/05/technology/anthropic-settlement-copyright-ai.html), etc. concerns as well. Let's not forget about those. Do we want to be a part of this? Maybe we do. Maybe we don't.

Just food for thought.

Reviewer 11: I'll add my own thoughts here, as a long-time NixOS user that is likely to jump ship if any GenAI contributions will be let in.

This policy is, to me, way too lax. I feel the bare minimum is something like what Rust is doing in https://github.com/rust-lang/rust-forge/pull/1040.

Personally, I am willing to compromise with using it to look things up, or to use it for reviews. But that is the furthest I would be ok with letting it in.

GenAI technology is fundamentally unethical, and I feel it is important to make a stance against it.

Reviewer 12: The status quo is that LLM-generated contributions are already being let in to Nixpkgs, silently, with no attribution, because there are (with the exception of the code of conduct) zero rules governing their use.  As we said in the PR body, "We would like this policy to be seen as establishing consistent standards based on pre‐existing norms around automation and ruling out the most problematic cases, not as an endorsement by the project, nor necessarily as the final word on the matter."  It would be very difficult, and take a much longer time, for a near total ban on LLMs to gain widespread community acceptance, which would be necessary to have it enforced effectively.  (And there are also other unintended consequences to consider, like people maybe being more likely to use LLMs unattributed with such a policy.)

That's why we want to start with a more scoped policy that we hope will be broadly acceptable.  That doesn't preclude a stricter policy later, if consensus and legitimacy in the community can be achieved for it; it just means that we don't get stuck with no policy at all (de facto permitting unrestricted LLM use) unless and until we get to that point.

Reviewer 9: > It would be very difficult, and take a much longer time, for a near total ban on LLMs to gain widespread community acceptance

Which is why a moratorium is a reasonable start

EDIT: Meaning, a moratorium until we've figured out how to define "safe ways of using \"AI\"" if any, such as a definition of some sort of "slow mode" operation: https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3167076386

Reviewer 11: > The status quo is that LLM-generated contributions are already being let in to Nixpkgs, silently, with no attribution, because there are (with the exception of the code of conduct) zero rules governing their use.

There not being any rules should mean erring on the side of caution and banning them, in my opinion. Though, I do get what you're saying.

> That's why we want to start with a more scoped policy that we hope will be broadly acceptable.

I fear that this policy will provide legitimacy to anyone trying to get GenAI outputs into nixpkgs, and that after it is accepted, there will be much less of a rush to get a proper policy in place.

> (And there are also other unintended consequences to consider, like people maybe being more likely to use LLMs unattributed with such a policy.)

Yes, they could try to sneak GenAI contributions in, but I don't think this should change the stances of the project. This technology is unethical and I feel it is important to make a stand against it.

Enforcement can be done in obvious cases. Having a stance against it will at least make people know that GenAI is unacceptable, and that is the most important thing, to me.

I feel https://cyrneko.eu/ai-policy.html has good ideas regarding enforcement:
> These policy discussions always sound a bit like those in tournaments or speedrunning communities around rules.
> 
> There, a lot of the same problems crop up in regards to cheating, just that here we're not talking about "cheating" in the traditional sense of course.
> 
> They usually handle it pretty simply:
> 
>  -  There is a team that checks anything submitted against the rules
>  -  If someone is found to violate the rules, it gets escalated and becomes a discussion with rulemakers and such
>  -  If the person is found to have cheated, they are banned from competing.
> 
> This applies pretty nicely to software projects as well, of course substituting some of the terminology for fitting ones for software development.
>
> If someone lies about the use of AI, or license compatibility despite set-in-stone clear policy around such things, they should be banned from contributing. Either permanently or temporarily.
>
> Just like with cheaters in games, sometimes someone will make it through successfully. But, more often than not, these people eventually also get exposed, and then swiftly banned from competing.

Reviewer 10: Here is a perfect example of a contributor just feeding a reviewer's comments back into an LLM: https://github.com/NixOS/nixpkgs/pull/514745

Reviewer 13: While I’m sympathetic towards this position (not accepting LLM-based contributions), I don’t think starting with a ban is politically viable in Nixpkgs. I would prefer to have some policy, which can be revised, than continue having no policy due to the initial attempt being too controversial to move forward.

Author: I really appreciate you taking the time to give such a thoughtful comment, @ethancedwards8; thank you. I’ll try to do justice to it and the other matters raised in this thread in response.

I want to say, first off, that it is absolutely the intention of this policy to **fully prohibit vibe coding**, in the sense of getting an LLM‐based AI tool to do all the work without review and submitting it without understanding or learning anything, and that we also completely agree that **acting as a proxy passing reviews unthinkingly back and forth to LLMs is not acceptable**. Those get to the heart of [“extractive contributions”](https://llvm.org/docs/AIToolPolicy.html#extractive-contributions), and immediately mitigating those by ruling out cases that we believe are obviously unsuitable for FOSS contributions is our primary objective with this policy.

We agree that the social contract of genuine interaction and learning is important, and a significant part of what we want to address here. While disclosure is the foundation of the policy, I think the rules around accountability are the critical ones. I would say the core part of the policy is this:

> Everyone who submits a contribution to Nixpkgs is responsible for it, regardless of the use of automated tooling. Before submission, they are expected to establish a reasonable level of understanding of the contribution and belief in its correctness. Contributors are expected to be able to answer questions about their contribution and respond to feedback appropriately.

Someone who has vibe‐coded a package without learning anything about Nixpkgs won’t have reached a reasonable level of understanding, won’t be able to perform the required review of the LLM output, and won’t be able to appropriately answer questions about it or respond to feedback; it’s true that these are somewhat subjective requirements to judge, but we thought they were important enough to include anyway.

The case of being a proxy ferrying reviews back and forth to the same tool that produced a contribution in the first place is why we’ve included PR reviews and comments as covered contributions with the same expectation of disclosure and responsibility; the requirements to review and understand what you’re submitting and to be able to handle questions and feedback about it appropriately apply equally to further communication on a PR as they do to the original submission, and are intended to require people to genuinely engage and put work into their communication, rather than forwarding things on to an LLM that won’t learn for the next PR. (Overall, I would say that code should be treated as a form of communication in itself, hence the policy treating it uniformly with comments and reviews.)

That said, we ought to at least take a look at clarifying the wording here to make the requirements clearer, and this kind of dynamic is very much an area where we expect there may be the need for further revision and refinement – we don’t think that the initial rules here necessarily guarantee that appropriate oversight and engagement is taking place, and it’s certainly not the case that the policy is intended as an endorsement of every contribution that meets its minimum requirements. 

---

On ethics: We by no means want to dismiss the social or ethical concerns people have here. There are some things that we can easily near‐unanimously agree are ethically unacceptable (e.g. harassment, submitting malicious code). The hard cases, of course, are where there’s significant division. For an example, there are many strong and well‐reasoned arguments that non‐Free software is unethical, but they’re far from universally accepted. We have community members who prefer to absolutely minimize their exposure to non‐Free software, community members who are completely fine with it, and community members who don’t have a strong opinion one way or the other or even feel conflicted about it (as a `lib/licenses.nix` maintainer who is not infrequently being a FOSS pedant on PRs, and a member of the Darwin team who daily drives Nix on macOS, I am familiar with this dissonance).

We have an essentially pragmatic policy that splits the difference as a result: community values encode our [preference for FOSS, but support for users’ need for non‐Free software](https://nixos.org/governance/#community-values-choice) explicitly; we allow arbitrary non‐Free packages, but we ensure that they are appropriately labelled, and only build packages with FOSS licences on Hydra.

That seems to work quite well in practice for our diverse, pluralistic community, but it’s not a fully settled issue (e.g. recurring discussions about building non‐Free software on Hydra), and some will consider it an unacceptable ethical compromise and be unwilling to contribute to or use Nixpkgs as a result.

I don’t mean to trivialize this or say that a stronger stance would inherently be unwarranted; GNU considers it a serious enough ethical issue that projects like Guix don’t package or recommend non‐Free software. But I think it would be fair to say that there’s no single “Nixpkgs stance on non‐Free software” as a result; our policies and norms in that regard, as with many others, are there to ensure that people with different views on the topic can work together while respecting each other’s positions and boundaries. (Which, I realize, may not be possible or desirable for every matter.)

There are, of course, many differences between non‐Free software and generative AI; I only mean to analogize them here insofar as they are both things relevant to Nixpkgs that many people consider deeply ethically fraught. One of the biggest differences is that non‐Free software has been around for a very long time; it’s more conceivable that the community’s views on generative AI will converge over time than for an issue that has existed since the 1980s.

---

We know that no matter what policy we start with, it wouldn’t be many people’s ideal, but we think that the vast majority of people will prefer some policy over none, and will agree that the things the proposed policy rules out are harmful to Nixpkgs. We want to approach this iteratively by starting with a baseline that rules out the cases that are most clearly unacceptable and ensures people are informed about the provenance of a contribution before engaging with it, but we don’t intend to set an immutable policy and then wash our hands of any problems that remain once it’s in place.

A lack of consensus doesn’t necessarily mean that action shouldn’t be taken; as we said in the original PR message, we are open to making judgement calls for the good of Nixpkgs as necessary, but we hope that setting basic principles for how people are expected to engage will help establish a foundation for further consensus and community discussion here, and that the experience with the resulting dynamics and enforcement will help us when considering the potential for further changes to address subtler or more controversial cases.

Reviewer 14: > We have an essentially pragmatic policy that splits the difference as a result: community values encode our [preference for FOSS, but support for users’ need for non‐Free software](https://nixos.org/governance/#community-values-choice) explicitly; we allow arbitrary non‐Free packages, but we ensure that they are appropriately labelled, and only build packages with FOSS licences on Hydra.

This is a false equivalence. We _package_ non-free software, but nixpkgs itself is free. Likewise, we _package_ some LLM-generated software, and here we are deciding whether to allow LLM output in nixpkgs. I think the equivalence should be between non-free code _in_ nixpkgs and LLM-generated code _in_ nixpkgs.

> But I think it would be fair to say that there’s no single “Nixpkgs stance on non‐Free software”

Sorry if this comes off as pedantic, but regarding non-free code _in nixpkgs_, the single stance exists [in the license](https://github.com/NixOS/nixpkgs/blob/master/COPYING).

If we are to copy our free-software policy, and agree that LLMs and their output are at least as problematic as non-free software, I think we should ban LLM output altogether, while keeping packages for LLM-generated projects to support the users' needs[^1].

[^1]: And ideally label them as LLM-generated, though I understand doing this accurately may be unfeasible at the moment.

Reviewer 15: > If we are to copy our free-software policy, and agree that LLMs and their output are at least as problematic as non-free software

«if we agree» evaluates to false over Nixpkgs, regardless of what goes afterwards.

Author: > This is a false equivalence. We _package_ non-free software, but nixpkgs itself is free. Likewise, we _package_ some LLM-generated software, and here we are deciding whether to allow LLM output in nixpkgs. I think the equivalence should be between non-free code _in_ nixpkgs and LLM-generated code _in_ nixpkgs.

Apologies; I didn’t intend to paint an equivalence that direct. As I said, the parallel I was attempting to draw is that “they are both things relevant to Nixpkgs that many people consider deeply ethically fraught”.

The intended analogy is between “using LLM‐based automation as part of Nixpkgs development to any degree” and “packaging/using non‐Free software to any degree”, not “allowing Nixpkgs itself to be non‐Free”. As I said, there are of course differences between the two cases even then – and those differences can factor into how they should be handled – but I think it is fair to say that they are both issues that many people reasonably consider to have a strong ethical dimension, and that in the case of packaging non‐Free software our long‐standing policy is essentially a pragmatic compromise between differing views (and certainly not a zero‐cost one – my understanding is that it was a factor in the creation of Guix, and there are also people who are unhappy with Hydra not building non‐Free software). I should probably have tried to find a less confusing metaphor, though.

(There is, of course, **very** strong consensus that Nixpkgs itself should remain under a FOSS licence (though there has been some preference expressed in the past for a potential relicensing under a licence with stronger requirements, to the point of coming up during SC elections, and I believe a small fraction of proposals there have gone beyond what’s widely accepted as FOSS, so it’s not completely unanimous). I don’t think it’s the case, however, that that consensus is because of community agreement on the ethical position of non‐Free software; it’s because people who have an ethical commitment to FOSS and people who have a utilitarian desire for a community‐developed FOSS Nixpkgs have common cause.)

> # Automation/AI policy

Reviewer 15: Would it be a good idea to insert something like «Optionality» section? In the direction that any automation that accepts unpredictability not coming from the very task statement (like if we package something compiled we have to live with the compiler's optimiser's quirks…) should not, under current version of the policy, be an expected part of the workflow?

Given that the policy will quite probably need to be updated either based on ethical negotiations or on legal clarifications, it would be prudent to make sure it is always about what contributions get accepted, not about suddenly losing load-bearing workflows touching all contributions.

(And also, S3 and GitHub is enough of dependencies on evil oligopolies for now, maybe we could skip adding another one)

Author: Can you expand on what you mean here, in terms of what rule it would be setting for contributions to follow? While there are already some subjective parts of the policy, I think it’s best for us to err on the side of trying to set rules only about things there’s some degree of “fact‐of‐the‐matter” about whether a contribution (or at least a contributor’s activity as a whole) meets; “expected part of the workflow” feels like it’s somewhat far afield of this, in a similar way to https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175202480.

Reviewer 15: Something in the direction of:

> Contributions that add automation to widely-used workflows, e.g. by adding CI checks, should only add automation that can be reasonably expected to be more or less deterministic and predictable, and to be developed towards improvent in determinism and predictability.

Reviewer 16: Is the hedging with "more or less" even needed? CI should be completely reproducible from one run to another. It sounds like a good time to make it explicit. While I'm fine with genAI use overall, this is one of the places I'd be happy to completely ban the idea.

Reviewer 15: The problem is, I can probably get a local LLM technically speaking «reproducible» with t=0 and fixed seed (but this is still not enough of predictability to go into CI), but ofBorg running the tests has the risk of tests being timing-sensitive in annoying ways, so not «completely reproducible». And, as I said above, once there is patching and also optimising compilers, «completely predictable» is also slightly in question, although not by much.

Author: Thanks, that’s clearer. It seems like this is probably something we don’t need to write down a strict guideline about, as it’s already an area with active maintenance and scrutiny and it seems like there’s not sufficient disagreement about it for local consensus to be insufficient?

Reviewer 15: From my observations, I would rate the addition of linting into CI as chaotic and subject to interactions of a growing number of maintainer teams of specific linters. An LLM-based linter will probably be eventually blocked anyway, but with a risk of unnecessary levels of heating up the discussion. Closing in accordance to the existing policy would have some benefits in cutting out the flaming.

A secondary use — to strengthen the signal that LLM output is untrustable on its own — might be considered a bug or a feature.

> # Automation/AI policy

Reviewer 15: A question on retroactive applicability of sorts: should we set an expectation that a contributor that remains active is expected to answer (roughly, not everyone remembers sufficient details and that's OK) about automation use in pre-policy contributions?

This might become relevant for preparing for eventuality where e.g. if France or Italy (choice of countries is not random) gets a regulation with a wider interpretation of derived-work-via-LLM-training-and-use.

Author: This seems broadly in line with expecting contributors who are still around to answer questions about the licensing/copyright status of code they’ve previously submitted in general, right? I think that it’s something we’d reasonably expect people to do and that it would not be very polite to arbitrarily refuse to do so if a concern comes up, but since the failure case of “they’re just not around in the project any more” is always present I’m not sure if it needs explicit policy wording beyond our general social norms (“Focusing on what is best for the community” per the CoC, and so on).

Reviewer 15: Strictly speaking we do not know exactly (yet) the extent — per-jurisdiction — of questions about licensing status, but maybe we should already collect the data where feasible. 

But as long as your choice is deliberate here, fair enough, thanks for consideration!

> Every contribution to Nixpkgs and related development venues, including code, documentation, and communication on GitHub and Matrix, must have a **responsible person in the loop** who is accountable for that contribution and reviews it before submission, and must **transparently disclose** any non‐trivial use of automation to produce it, including but not limited to LLM‐based AI tools.

Reviewer 15: ```suggestion
Every contribution to Nixpkgs and related development venues, including code, documentation, and communication on GitHub, Discourse, and Matrix, must have a **responsible person in the loop** who is accountable for that contribution and reviews it before submission, and must **transparently disclose** any non‐trivial use of automation to produce it, including but not limited to LLM‐based AI tools.
```

Reviewer 17: Do we consider Discourse a nixpkgs development venue? (honest question; I see it more as a support venue most of the time)

Reviewer 4: > Category: Development
> Development discussion for Nix, NixOS, Nixpkgs, NixOps, RFCs, etc.

Reviewer: Not in entirety, but the specific threads/topics/spaces. Same with Matrix, only development chats (`#dev`, `#ci`, `#hydra` and so on). I'd suggest specifying that moment:
```suggestion
Every contribution to Nixpkgs and related development venues, including code, documentation, and communication on GitHub and in development-related spaces on Matrix and Discourse, must have a **responsible person in the loop** who is accountable for that contribution and reviews it before submission, and must **transparently disclose** any non‐trivial use of automation to produce it, including but not limited to LLM‐based AI tools.
```

Author: The reason Discourse wasn’t included here originally is that it covers a range of topics outside our jurisdiction, like Nix itself. On GitHub and Matrix, the categorization is distinct enough that it’s fairly unambiguous which venues are related to Nixpkgs and therefore in scope; on Discourse, a lot of topics are just in the “Development” category with no further categorization.

I think it would be a good idea in itself to include Nixpkgs‐related topics on Discourse, but the discontinuity between standards in one topic to the next depending on its subject (and the ambiguity when that subject shifts) may be surprising, compared to GitHub and Matrix where the subject of a venue is more static. Happy to hear people’s thoughts about this.

@acid-bong I think your suggestion is redundant: “Every contribution to Nixpkgs and related development venues, including … communication on GitHub and Matrix”? Note that even for GitHub, related development venues does not include e.g. the Nix and Hydra repositories, because they’re outside of the Nixpkgs core team’s authority.

(Of course, we’d be happy to see this work grow into a more widely‐applicable policy beyond Nixpkgs in future!)

> Every contribution to Nixpkgs and related development venues, including code, documentation, and communication on GitHub and Matrix, must have a **responsible person in the loop** who is accountable for that contribution and reviews it before submission, and must **transparently disclose** any non‐trivial use of automation to produce it, including but not limited to LLM‐based AI tools.

Reviewer 18: I would strongly consider an exception to disallow LLM generated pull request comments that are in response to reviewer feedback. It seems rude, and is a great way to get a reviewer to never look at your PR again. This should not apply to the translation or speech to text accessibility use cases.

Reviewer 19: I've seen (but can't remember where) policies that say that all text (PR bodies, comments etc) must be human. I like that approach personally; both because I don't enjoy talking to machines in the 'human conversation' part of the repository but it also gives a fairly clear-cut reason to dismiss no-human-in-the-loop contributions, where the author doesn't show any engagement (and often becomes a proxy to their coding agent).

Additionally, if someone puts their contribution in their own words, they might actually review what they are submitting to understand it and stand a chance of spotting mistakes/suspicious things/... before a reviewer has to do that for them.

I see somewhat frequently well-meaning people [on other repos] using these tools and (innocently!) submitting something that they glanced at, thought it looked plausible, but later dawned on them that it was the LLM being silly.

Honestly I would also advocate for code comments being human-written only; I've also experienced submissions where the code comments were LLM-generated, had the wrong rationale/no understanding of the real context and the human didn't realise until the reviewer probed them on it...
Harder to guarantee though.

Reviewer 20: It's ok to set some expectations even if they are not exactly enforceable individually, because they're still informative to the contributor and still moderately useful to the reviewer as well.

Author: We absolutely agree that we’ve seen some very unfortunate dynamics here; I’ve talked about this in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175207606.

> # Automation/AI policy
> 
> Every contribution to Nixpkgs and related development venues, including code, documentation, and communication on GitHub and Matrix, must have a **responsible person in the loop** who is accountable for that contribution and reviews it before submission, and must **transparently disclose** any non‐trivial use of automation to produce it, including but not limited to LLM‐based AI tools.
> 
> The following sections give more detail.

Reviewer 9: Rendered: https://md.someonex.net/s/Nzr2olhfk8#

```suggestion
# Automation and "AI"

## Care and respect

All text intended for a human reader must be human-authored.

## Policy

Every contribution to Nixpkgs and related development venues, including code, documentation, and communication on GitHub and Matrix, must have a **responsible person in the loop** ("Contributor"). This person is accountable for the contribution, reviews it prior to submission, and must **transparently disclose** any non‐trivial use of automation involved in producing it, including but not limited to "AI" and (L)LM-based tools.
```

Reviewer 15: > complexity, poor inspectability, poor predictability, &c

Speaking of what is more code and what is more data, lockfiles sometimes get committed, and I think some ecosystems use SAT-solvers, which are complex, have high input sensitivity / low predictability of choice between technically allowable solutions, and also have once been called AI.

> Yes, Nixpkgs is "applied maths".

Most of Nixpkgs _by impact_, possibly. Most of Nixpkgs by churn is at best computational linguistics.

> ..."working software" here meaning "working in the long(er) term", i.e. "bootstrapable, maintainable, inspectable", of course.

Honestly, for periphery packages often no.

There are good upstreams, where expressions do not require much maintenance beyond versions and maybe dependency lists, and there are bad upstreams where expressions need periodic overhaul as careful as you ever are. There are some medium upstreams where the situation is in the middle, but I am not sure they are the majority of our churn or anything.

> Whether we're talking about Nixpkgs-for-deploying-today, or Nixpkgs-for-reading-by-humans, or the Nixpkgs-for-data-analysis, all three are of most value when written by humans.

Nixpkgs-for-deploying-today is a thing that is of most value when it has coverage. Some things need to be well-designed for that (i.e. cross-compilation is a large increase in coverage). Some things just need to be there at all, more things existing better than fewer but higher quality. 

Oh, and we have already plenty of human-intent-destroying policies as well as policy-enforced falsehoods if we look at the case of humans reading expressions.

Reviewer 9: > and I think some ecosystems use SAT-solvers, which are complex, have high input sensitivity / low predictability

👍🏿 👍🏿 👍🏿

## Scope

Any use of automated tools to generate non‐trivial amounts of output as part of a contribution, in whole or in part, verbatim or edited, is covered by this policy, except as listed in the Exemptions section.

> Both LLM‐based AI tools and hand‐written automation are covered.

Reviewer 21: Does this include deterministic template-generating tools such as [nix-init](https://github.com/nix-community/nix-init)?

Reviewer 22: TBH, disclosing use of nix-init is IMO beneficial to the reviewer.

Reviewer 10: Agreed. Also fwiw I think nix init is still configured to use some anachronisms like rev and sha256 (at least last time I used it)

Reviewer 23: > TBH, disclosing use of nix-init is IMO beneficial to the reviewer.

Agreed too. And that could be automated for the `--commit` flag, where adding a `Assisted-by: nix-init` would make sense

> I think nix init is still configured to use some anachronisms like rev and sha256 (at least last time I used it)

that is fixed now

Reviewer 24: > > I think nix init is still configured to use some anachronisms like rev and sha256 (at least last time I used it)
> 
> that is fixed now

Sure, that individual case has been fixed, but that's not the point; the point is that `nix-init` will always lack in adherence to nixpkgs standards, and I could name a couple of other basic issues that have not been addressed for literal years.

Reviewer: rafamadriz/friendly-snippets also hasn't been updated in 2 years: https://github.com/rafamadriz/friendly-snippets/blob/6cd7280adead7f586db6fccbd15d2cac7e2188b9/snippets/nix.json

Reviewer 25: opened https://github.com/nix-community/nix-init/pull/784 to add the assisted-by trailer to `nix-init --commit`

> the point is that `nix-init` will always lack in adherence to nixpkgs standards, and I could name a couple of other basic issues that have not been addressed for literal years.

nix-init was barely maintained for a while, I've since picked up the maintenance again and fixed most of the outdatedness. one of my goals for nix-init is to closely follow nixpkgs conventions, please open an issue upstream for any issue you are still running into, so I can take a look at them @eclairevoyant

Author: I would expect that `nix-init` probably falls under:

> Use of standard community automation is exempt, such as `nix-update`, the official Nixpkgs CI bots, the @r-ryantm update bot, and the Nixpkgs security tracker bot.

“Standard community automation” is admittedly fairly vague, but it seems better for it to be interpreted reasonably broadly rather than treating things that have previously been widely accepted as violations.

I agree that it would generally be good for automated tools of all kinds to disclose themselves in commit messages. Still, I don’t think we’d want to consider it a problem for people to run `nix-update` without disclosure, especially as these tools don’t always create their own commit messages.

> Contributions include code and documentation in commits, commit messages, pull request summaries and reviews, issue and vulnerability reports, GitHub comments, and Matrix messages.

Reviewer 15: ```suggestion
Contributions include code and documentation in commits, commit messages, pull request summaries and reviews, issue and vulnerability reports, GitHub comments, and messages in other official community spaces.
```

Author: (I think this is the same case as https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175198295; let me know if there’s nuance I’m missing.)

> The covered venues are the GitHub repositories for Nixpkgs and [related projects](https://github.com/orgs/NixOS/teams/nixpkgs-core/repositories) under the jurisdiction of the Nixpkgs core team, and Matrix rooms that are focused on development of those projects.

Reviewer 26: Sorry if somewhat OT, not sure where to put. I would try discourse but there's more SealionGPT over there than in here. From the PR message:

> The [Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) has a clause against “Wasting other people’s time with low quality contributions, including but not limited to LLM and bot spam”, but does not define this in detail

> After discussion with the community team bootstrap group and within the Nixpkgs core team, we’ve agreed on submitting this proposed policy

Is there a CoC-side policy forthcoming in addition to this contributing.md diff? I ask because it feels like the limited scope here leaves a massive policy vacuum in moderation, and IMO chatbots speaking for people in discussion is more socially corrosive than garbage code, whether it's PR comments or forum posts.

Even a placeholder "bots should not speak for you outside of properly disclosed translation/dictation" would enable reporting and removing the most egregious cases.

Yes I know the forum upheaval context and I'm grateful to people picking up the pieces. But it sucks having to second-guess every new person's first post under the status quo.

Reviewer 26: And to provide a constructive example, I think the Rust project gets it right on the discussion/comments side in a manner that would get to consensus more easily here than code or docs:

> ❌ Banned
> The following are banned.
> - Comments from a personal user account that are originally authored by an LLM. 
>   -    ℹ️ This also applies to issue bodies and PR descriptions.
>   -    ℹ️ See also "machine-translation" in ⚠️ below.

> ⚠️ Allowed with caveats

> -    Using machine-translation (e.g. Google Translate) from your native language without posting your original message. Doing so can introduce new miscommunications that weren't there originally, and prevents someone who speaks the language from providing a better translation.
>      - ℹ️ Posting both your original message and the translated version is always ok, but you must still disclose that machine-translation was used.

> All of these **must** disclose that an LLM was used.

Reviewer 15: This is even less restrictive than it looks like, because you can quote tool output with proper explanation what you are quoting (we are quoting build failures all the time in the issues). But yeah, it's better that first-person speech is always coming from a person.

Author: FWIW, this policy does already cover PR comments; I think https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175207606 is relevant to your points here. Discourse topics are not covered at present, but topics about Nixpkgs development perhaps could be; see https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175198295.

Given our jurisdiction, I would say that interpretation and enforcement of the CoC clause more generally, and setting policy that applies beyond Nixpkgs, is a matter for moderation and the Steering Committee. As I said in the other comment, we’d of course be very happy to see our work here grow into a more widely‐applicable policy in future.

Reviewer 26: Replacing a longer frustrated response: Thanks for your work here. Grateful despite the gnashing of teeth and two thumbs up on the new policy overall.

> Given our jurisdiction, I would say that interpretation and enforcement of the CoC clause more generally, and setting policy that applies beyond Nixpkgs, is a matter for moderation and the Steering Committee.

> After discussion with the community team bootstrap group and within the Nixpkgs core team, we’ve agreed on submitting this proposed policy for community review.

To read this as optimistically as possible, it sounds like you're saying this policy, despite being the output of your collaboration with the (pre-)community team, is _not in lieu of_ general moderation changes, and other community policy improvements are forthcoming.

Is that as naïve as it sounds on a second read? I hope not.

Author: Members of the community team bootstrap group were enthusiastic about having a policy that goes beyond Nixpkgs in our discussion. I believe they’re still getting onboarded, though, so this is just a matter of me not knowing what concrete plans are in the pipeline and not wanting to speak for them rather than me trying to be deliberately vague, I’m sorry 😅

(For clarity, we started discussing this in the Nixpkgs core team before the community team was announced, so we already had a draft policy by the time we established a line of contact with them to present it for feedback, and which was then revised in response to that discussion.)

Reviewer 23: Speaking for the boopstrap team, I can confirm that we would like to see this work by the nixpkgs core team extended to the whole org, yes.

## Accountability

Everyone who submits a contribution to Nixpkgs is responsible for it, regardless of the use of automated tooling.

> Before submission, they are expected to establish a reasonable level of understanding of the contribution and belief in its correctness.

Reviewer 27: Taken too literally, "belief in its correctness" would seem to prohibit e.g. submitting a PR that is known to be a regression for feedback. Maybe:

```suggestion
Before submission, they are expected to establish a reasonable level of understanding of the contribution and (unless they disclose otherwise) belief in its correctness.
```
?

Author: Perhaps this is covered by this existing exemption?:

> * Use of automation in a contribution clearly marked as not being ready for review (e.g. a draft pull request) is exempt from the requirement for full self‐review, as long as some amount of review has been done and it is expected that the requirements will be met by the time it is marked as ready. This does not waive any other requirement.

Edit: Ah, I guess “ready for review” here is too strong. Maybe just adjusting that to “ready for merge”?

Contributors are expected to be able to answer questions about their contribution and respond to feedback appropriately.
A contributor submitting a contribution intended for inclusion in Nixpkgs is also responsible for ensuring that it is [appropriately licensed](https://github.com/NixOS/nixpkgs/blob/master/COPYING) and credited, and not encumbered by any incompatible copyright.

> When output from automated tooling is used in contributions, a contributor must establish confidence in that output.

Reviewer 7: One thing I would like to request, though I'm not sure where it'd go here, is a sort of rule that:

> Contributions from tools should at the very least reduce toil for future maintainers.

The idea there being that if you contribute via LLM or similar for packages or services, we expect to see:

* Addition of tests (if they were missing)
* Addition of update scripts
* Addition of (as much as possible) well-documented and well-typed program and service options (instead of, say, the not-uncommon practice of "here's a config string we'll turn into a file, good luck").

([nix-facts](https://github.com/crertel/nix-facts) is a tool you can use to try and see how many packages are missing tests or update scripts, though it's in "beta")

Basically, if we use LLMs, we should be using them not only for our own convenience but to achieve code quality and ease-of-maintenance that is currently not there.

Reviewer 6: I think that this is something valuable to keep in mind as a principle, but I think it's a little bit too subject to interpretation to make a hard and fast guideline.

Reviewer 7: I think a hard-and-fast rule of "if you're adding a package/service using an LLM, we better see tests and update scripts" is a pretty objective measure--either the tests and update scripts are there, or they aren't.

But, I'm not sure that belongs in *this* doc, as such.

Reviewer 28: "leave the codebase better than you found it" is a good guideline, but would be indeed subjective to be a hard rule.

May be worth a mention *somewhere* in the contribution guidelines though. Explicitly, because even though it's common-sense for many long-time maintainers, I feel like newcomers may benefit from a reminder. And I feel the positive sentiment as in "try to strive for this ideal" would offset the "don't do this, don't do that" of hard-and-fast rules aimed at low-effort contributions mentally, encouraging good contributions instead of just discouraging bad ones.

Author: We would definitely like to encourage contributors to raise our technical standards! But I agree with the other replies that this is a more general principle, and would best go in other contributing documentation, as it seems particularly important for this policy to stay focused; any subjectivity is more of a risk here than elsewhere, and I think it already risks being too verbose in the effort to avoid that.

This can be achieved by establishing confidence in the correctness of the tooling’s logic, manual review of the included output, or using further automation to verify the output (e.g. programmatically checking whether a refactor avoids causing rebuilds).
As the inner workings of LLM‐based AI tools cannot be sufficiently understood at present, only the latter two options are available when those are used.

> When automation is used to verify output, the verification tooling itself must be disclosed and reviewed in line with this policy.

Reviewer 14: I challenge the assumption that the third option can successfully establish confidence in the output of an LLM.

LLMs are fundamentally unpredictable. For every verification automation you can devise there likely is an output that works around it. For example, if you "programmatically check whether a refactor avoids causing rebuilds", a LLM could insert malicious code hidden behind a flag that is disabled by default.

This is not hypothetical. There have been numerous reports of LLMs working around restrictions imposed by the user, for example by removing tests that don't pass or even going as far as escalating privileges.

Reviewer 9: +1, was going to bring this up as well, but got stuck formulating a `suggestion` block

Neither a manual post-hoc review of the outputs, nor a post-hoc harness are necessarily sufficient in themselves. I'm more skeptical about the manual review than about using a harness, because of the pace mismatch between the human operator and the model, and because of how human attention works. What I'd require from people experimenting with LLM usage in Nixpkgs is a form of "slow mode": https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3167076386.

Author: I agree that this method is not wholly reliable, but I’m not sure that reading the diff is either, in the limit of [underhanded code](https://en.wikipedia.org/wiki/Underhanded_C_Contest). I would say that this is what we have reviewers for – we can’t assume that every contribution we get is correct or even non‐malicious, and so we try to have a second pair of eyes on changes, while accepting that that won’t necessarily catch everything. The keyword is “second”, though; we can at least establish a baseline of the contributor having done something to gain confidence in it, if they’re acting in good faith, regardless of whether automation was used or not.

These considerations apply to treewide refactors in general, many of which are too big for anyone to practically deeply and exhaustively review their entire diff by hand, which was our motivation for mentioning automated validation here – we’ve had both successful and unsuccessful automated treewides in the past that has correlated strongly with the amount of mechanical validation. I’ve merged at least one large partially‐automated, partially‐manual treewide that would have been intractable to write or verify fully by hand, based largely on automated review (https://github.com/NixOS/nixpkgs/pull/398707#issuecomment-2816914489).

Since this criteria is just covering what needs to be done before *proposing* a change, and we explicitly state that “Automation without any manual review must not be used as the sole arbiter of whether to merge a change” and “Everyone who submits a contribution to Nixpkgs is responsible for it”, I believe it doesn’t have to be perfect, and for certain types of change I believe automatic checking does better than manual review; I won’t say that a refactor being labelled as causing zero rebuilds is proof that it’s correct, but I think it’s at least reasonably strong evidence for the expected level of risk, and there’s certainly been many cases where I intended for a refactor to be zero rebuilds but found out it wasn’t and had to correct it.

Reviewer 14: Thank you for your thoughtful replies. My understanding is that you agree that method 3 is not reliable when applied to LLMs[^1], and that large changes are impossible to review (method 2). If so, it follows that LLM use should be forbidden at least for large changes.

I suggest:

```diff
When output from automated tooling is used in contributions, a contributor must establish confidence in that output.
-This can be achieved by establishing confidence in the correctness of the tooling’s logic, manual review of the included output, or using further automation to verify the output (e.g. programmatically checking whether a refactor avoids causing rebuilds).
+This can be achieved by establishing confidence in the correctness of the tooling’s logic or manual review of the included output, and can be supplemented by using further automation to verify the output (e.g. programmatically checking whether a refactor avoids causing rebuilds).
-As the inner workings of LLM‐based AI tools cannot be sufficiently understood at present, only the latter two options are available when those are used.
+As the inner workings of LLM‐based AI tools cannot be sufficiently understood at present, only the second option is available when those are used.
+It follows that LLM use for changes that are too large to be reviewed (such as treewide refactors) is forbidden.
When automation is used to verify output, the verification tooling itself must be disclosed and reviewed in line with this policy.
```
[^1]: Or in general unless either method 1 is valid as well, or the author of the change is trusted. Hence "supplemented by" in my suggestion.

Reviewer 15: There are some large changes where you can design formal (held-out — not available to the LLM, and with a strong syntactic component — no new conditionals etc.) checks that are easily reviewable and sufficient to check the change. Of course then the checks would need to be shared and also reviewed by other people.

General impossibility is there, but particular cases can be verifiable if diff shape is constrained enough.

Author: > My understanding is that you agree that method 3 is not reliable when applied to LLMs[1](#user-content-fn-1-c56ca841d48d3967063dc9bf5d70c866),

I would say rather that none of the methods are reliable in the general case for any automation, and that we compensate for this for major or treewide changes by expecting both that contributors establish some reasonable confidence in them, and that at least one separate reviewer does so independently.

I believe that for large treewides, automated checking has generally been more reliable than manual diff inspection, and there’s a difference between the [Goodhart’s law](https://en.wikipedia.org/wiki/Goodhart%27s_law) effect from “here’s a metric; optimize for it” vs. an out‐of‐the‐loop post‐facto mechanical check. I think that, as @7c6f434c says, sufficiently rigid checking can mitigate the risk significantly, and that we are exposed to it from any automated treewide – e.g. we’ve had treewides based on simple textual replacement that were merged due to being zero rebuilds but that broke the NixOS modules they touched that aren’t factored into rebuild counts. I don’t think refusing to do any treewides too large for manual authorship is the solution, though; we just need to be rigorous about them to ensure they improve quality rather than risking harming it.

None of this is to say that we should be merging a treewide without appropriate scrutiny of the methods used to produce it and without careful consideration of the best way to gain independent sources of confidence in it; anyone merging any treewide of course ought to at minimum spot‐check and skim the diff rather than relying solely on a mechanical check, but “Automation without any manual review must not be used as the sole arbiter of whether to merge a change” covers that. The fact that the use of verification code to establish confidence in a change itself requires review and disclosure means that reviewers also have an opportunity to give independent review of the pre‐submission testing methodology.

The intention here is to ensure that any automated change undergoes pre‐screening from a contributor before a reviewer has to consider it, not to mandate that every PR that gets opened is in an immediately mergeable state without further scrutiny.

Reviewer 14: > There are some large changes where you can design formal (held-out — not available to the LLM, and with a strong syntactic component — no new conditionals etc.) checks that are easily reviewable and sufficient to check the change.

@7c6f434c Could you make a practical example? I genuinely struggle to imagine a situation in which writing such a check is less complex than writing a deterministic tool to perform the change in the first place.

Reviewer 15: If one splits a package into -libs and -daemons, and in the treewide part only replacement of X with X-libs or X-daemons or both happens — this needs checking nixpkgs-review and tests (obviously), and it needs checking that indeed the diff has this restricted structure of replacing one thing with one of the three options, but writing a proper deterministic tool might be annoying.

Reviewer 9: > in the limit of [underhanded code](https://en.wikipedia.org/wiki/Underhanded_C_Contest).

I like the reference, maybe it's worth including in the policy as communicating the spirit...

> Before submission, they are expected to establish a reasonable level of understanding of the contribution and belief in its correctness.
> Contributors are expected to be able to answer questions about their contribution and respond to feedback appropriately.
> A contributor submitting a contribution intended for inclusion in Nixpkgs is also responsible for ensuring that it is [appropriately licensed](https://github.com/NixOS/nixpkgs/blob/master/COPYING) and credited, and not encumbered by any incompatible copyright.
> 
> When output from automated tooling is used in contributions, a contributor must establish confidence in that output.
> This can be achieved by establishing confidence in the correctness of the tooling’s logic, manual review of the included output, or using further automation to verify the output (e.g. programmatically checking whether a refactor avoids causing rebuilds).
> As the inner workings of LLM‐based AI tools cannot be sufficiently understood at present, only the latter two options are available when those are used.
> When automation is used to verify output, the verification tooling itself must be disclosed and reviewed in line with this policy.

Reviewer 9: Rendered: https://md.someonex.net/s/Nzr2olhfk8#Accountability

```suggestion
Contributor is responsible for their Contribution, regardless of the use of automation.
Before submission, they are expected to establish a reasonable understanding of the Contribution, and a belief in its correctness.
Contributors are expected to be able to answer questions about their Contribution, and respond to feedback appropriately.
Contributor submitting a contribution intended for inclusion in Nixpkgs asserts that it is correctly attributed, [licensed](https://github.com/NixOS/nixpkgs/blob/master/COPYING), and not encumbered by any incompatible copyright.

When an Output from automated tooling ("Tooling",) is used in a Contribution, Contributor must establish confidence in the Output.
This is achieved by establishing confidence in the Tooling, through manual Review of the included Output, or using further automation (a.k.a. "Harness") to verify the Output. A Harness may be e.g. a test suite or a programmatic check, ensuring that a refactoring does not trigger rebuilds. The Harness itself is considered an automation tooling subject to this policy.

Due to the inherent complexity and opacity of "AI" and (L)LM-based tools, their Outputs must be reviewed manually and (or) verified using an automated Harness. Note that "AI"-based tools might generate Outputs faster and more complex, than is feasible for a human Review or for an automated Harness to achieve reasonable coverage of.
```

Reviewer 8: if you run a test it must be disclosed in the commit in `Assisted-By:` format? this wording sounds a bit broad to me.
gotta say i also feel a bit mixed on the term harness here maybe as that also seems used for coding agent client programs.

this wording also i think leaves room for a liberal reading that it could suffice to have LLM output verified by LLM (surely as 'further automation' LLMs might qualify as a 'Harness'), which i think can help but should not be sufficient.

Reviewer 12: I think it would be more effective to stick to focus on what specific changes in meaning you'd like to see (as you've done a great job at above — thank you) than to propose large rewrites like this, which are difficult for us to incorporate in a redraft when we also have to take into consideration everybody else's comments.  For this particular diff I think the legalistic language would make the policy quite a bit harder for people to understand.

> 

Reviewer 19: Not sure where to leave this thread as it's not entirely 100% on topic to other threads, but wondering if there should be distinctions between core code and leaf code like the package definitions, in terms of (principally) LLM-generating it and the impact it has on other people in terms of maintenance later on.

A few months ago (when I had a subscription as a trial) I used an LLM to package some software in NixOS. These packages, I'd already tried to write by hand a year or two ago, but gave up because of the amount of trial and error involved.
(Each build takes so long and if you make one tiny mistake you have to try again)
I don't have time to try that many permutations of a build so I gave up.

Well, LLMs changed that — I was able to pose the problem to an LLM and wake up in the morning with a working package.
The package code works (I can run the programme), it *makes sense* to me, though I didn't write any of it.
I have permuted it a bit to see exactly why each non-trivial part of the definition is needed, to get a better understanding, but I'll never be able to claim with a straight face that I *wrote* it.

Should I be allowed, encouraged or discouraged to submit this package for inclusion? (So far I have not, I want to be respectful)

Plus points:
- Other people have requested the package in an issue. It'd be actually helping people to run software they want to run.
- If my submission is harmful, it's only a package, at the leaf of the code graph, it's not really inconveniencing anyone, is it? (It feels like it'd be different if it's part of `nixpkgs.lib`)
- With my middling understanding of Nix and Nixpkgs norms, it seems OK (I'm not proposing submitting it blind). I'm not *clueless* about how to build software and about how Nix works.

Negative points:
- I don't want to force people to review this if they don't like the idea.
- Arguably I still don't have the full experience of having written this, so I'm winging it a bit. (But to an extent, if I was just bashing my head against the keyboard trying to package this, I'm not sure I'd be faring much better.)
- How does anyone else know I'm not clueless :)

Reviewer: No matter how close your package to leaves, it's not the only criterium. Popularity, despite being loose and, in case of Nixpkgs, unmeasurable, also adds to the importance.

Now to your points:
- your pros:
  1. not applicable, package requests stopped being accepted about a year ago
  2. depends on how many people use it and care about it and its Nix recipe
  3. you will eventually gain more experience in how Nix(pkgs) works, either by contributing to Nixpkgs or by maintaining your system/home config (or both)
- your cons:
  1. the policy in question already covers the acceptability
  2. (and 3.) it's always better to learn stuff yourself than pretend to know more than you actually know. say, you package doesn't include something one considers important (like systemd units or shell completions), they reach out to you, but you're clueless about that exact aspect because you vibe-coded it. would it be a good situation for you and them?

Reviewer 15: > not applicable, package requests stopped being accepted about a year ago

The fact that we (correctly) do not care about non-contributing user package needs enough to track them does not mean that their needs do not exist at all.

> how Nix(pkgs) works

And also how it doesn't…

> say, you package doesn't include something one considers important (like systemd units or shell completions), they reach out to you, but you're clueless about that exact aspect because you vibe-coded it. would it be a good situation for you and them?

I am willing to explicitly claim, from experience, that having written a package that skips some parts of upstream stuff doesn't help me answer questions «how to add feature X that I don't care about for my use case» any.

> it's always better to learn stuff yourself than pretend to know more than you actually know.

To be honest, if someone did some mutation testing of the resulting package (which is an exception not the rule, with or without LLMs) I am no longer sure that they understand the result worse than after blindly fixing yet another error until things seem to work and ship-whatever happens.

Maybe the evidence of such mutation testing could be requested, however — in particular, because the error snippets would explain the package better.

> `nixpkgs.lib`

And this is an important distinction: `lib` is a codebase with design and architecture, most packages are parts of a _database_ even it somewhat pretend to be code.

Reviewer: > I am willing to explicitly claim, from experience, that having written a package that skips some parts of upstream stuff doesn't help me answer questions «how to add feature X that I don't care about for my use case» any.

fair enough. i initially wanted to steer in a different direction, like "say your package doesn't build, but you don't know how to debug the breakage"

Reviewer 13: I don’t think there should be a distinction between core and leaf packages. It invites argument over how much or whether a package is core or leaf, and we should want contributors to take responsibility regardless. Ideally, someone who contributes a new package should also be willing to maintain it.

Reviewer 19: Thanks for the comments

> I don’t think there should be a distinction between core and leaf packages. It invites argument over how much or whether a package is core or leaf, and we should want contributors to take responsibility regardless. Ideally, someone who contributes a new package should also be willing to maintain it.

maybe core/leaf packages is the wrong term, I meant more like this distinction from @7c6f434c 

> And this is an important distinction: lib is a codebase with design and architecture, most packages are parts of a database even it somewhat pretend to be code.

I (personally) would not be remotely comfortable at all vibecoding/LLM-assisted-coding stuff into `lib`. That feels like real, critical code that touches a lot of other people and a lot of other packages.

And to be clear, I am not suggesting lobbing a package into Nixpkgs with no maintainer. I'm happy to maintain the package, just the initial process of writing that initial build script is so painful for some projects that it's a huge barrier to packaging, because the iteration time is so long.

If I fail (which I'm not sure changes depending on whether that initial setup was LLM-driven or not)? Nixpkgs as a project can remove 1 dir from the codebase, there's no painful migration involved, it's as if I never arrived at the scene.

I am not saying if this is right or wrong, curious for perspectives, that's all.

---

>  it's always better to learn stuff yourself than pretend to know more than you actually know. say, you package doesn't include something one considers important (like systemd units or shell completions), they reach out to you, but you're clueless about that exact aspect because you vibe-coded it. would it be a good situation for you and them

To be fair, it's not that I don't understand things, or am suggesting not learning, but some projects just have such nasty build scripts/systems.
My actual knowledge of fundamentals like systemd units, `.desktop` files, etc is good. My understanding of how Nix is structured is good.
My day-to-day practice is a little weak because I don't live and breathe Nix every day, but it's enough to get by for writing NixOS and home-manager configurations with basic abstractions, plus packaging the software that I author.

My experience of packaging *weird* projects with *weird* build scripts/environments/requirements and dependencies is arguably the limiting factor; but I will not be doing that enough to be able to materially 'learn' anything — if I've already tried to package something (and I genuinely mean tried) and I can't find enough reference material already in Nixpkgs and I still can't convince upstream's tools to work. To an extent it's sort of just busywork.

Reviewer 13: > I (personally) would not be remotely comfortable at all vibecoding/LLM-assisted-coding stuff into `lib`. That feels like real, critical code that touches a lot of other people and a lot of other packages.
> 
> And to be clear, I am not suggesting lobbing a package into Nixpkgs with no maintainer. I'm happy to maintain the package, just the initial process of writing that initial build script is so painful for some projects that it's a huge barrier to packaging, because the iteration time is so long.
> 
> If I fail (which I'm not sure changes depending on whether that initial setup was LLM-driven or not)? Nixpkgs as a project can remove 1 dir from the codebase, there's no painful migration involved, it's as if I never arrived at the scene.
> 
> I am not saying if this is right or wrong, curious for perspectives, that's all.

Thanks for the clarification, and my apologies for implying that you might be suggesting adding packages without maintaining them. Given the clarification, I still wouldn’t want to make a distinction. If we did, I’d want those exceptions enumerated (e.g., this path or that package have special rules), but I’d rather not.

My reasoning is that if the proposed policy isn’t good enough to keep the quality of code in those critical packages high enough, then it’s probably also not good enough for those peripheral packages. We ought to be tracking those problems regardless (to inform any changes needed to the policy), but things fall through the cracks sometimes.

Reviewer 15: > say your package doesn't build, but you don't know how to debug the breakage

… looking at the ignorelist of LibreOffice tests that a few people tried to fix and nobody succeeded…

Packages just differ in how much effort and competence gives you what, and there are levels of effort and competence where using an LLM could probably help you improve your understanding by skipping the teaches-you-nothing parts of busywork, and there are levels of involvement and competence where using an LLM gives you both a worse expression and worse understanding, and there are combinations where you can get a better package and worse understanding or a worse package and better understanding, depending on how you use the tools…

> a distinction between core and leaf packages

Nixpkgs is pretty bad with making things binary that have many more intermediate states, and here it is also a spectrum which gradually shifts the trade-offs. The very core packages (even data-entry-style ones) have huge amount of downstream fallout when some previously documented case is not correctly reasoned about, so yeah, a riskier idea to apply LLMs there, but also these are the packages that unlike most others might get multiple competent and thorough reviews for a change. Typical packages gets less than one _thorough_ review per change.

And also we should be better at recognising levels of maintainer involvement: a leaf package that works better than nothing is fine, actually; if it gets dependents and some user of a dependent package considers improving the original dependency, we should be better at giving information whether this user already cares about the package more than the original maintainer. In which case maintenance handover will probably boost the amount of human thought invested above the threshold where LLMs might break more than they fix. 

> My reasoning is that if the proposed policy isn’t good enough to keep the quality of code in those critical packages high enough, then it’s probably also not good enough for those peripheral packages.

Quality is not to put in front of quality throne, quality is to improve people's experience, and it costs effort, and for more peripheral packages availability at bad quality is often the best effort-benefit trade-off.

Reviewer 13: > Quality is not to put in front of quality throne, quality is to improve people's experience, and it costs effort, and for more peripheral packages availability at bad quality is often the best effort-benefit trade-off.

I’m using code quality as a proxy for determining how well this policy is working. Maybe it’s not the right way to measure it. Whatever that is, we should be able to make inferences based on how well it works in areas with a lot of scrutiny (i.e., “core” areas) to gain an intuition of how well it may working in areas with less.

Circling back to https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3168857895, I think it’s a problem if there’s a distinction between “core” and “leaf” (or “critical” and “peripheral” or whatever). This is less of a technical concern and more about how the policy relates to contributing to Nixpkgs. If there are different rules depending on where or what people contribute, it will be a source of friction (and even worse if those rules are unwritten).

Reviewer 15: Right now we have a spectrum of impact between packages, and a spectrum of likely impact of changes (there are some changes to high-impact packages that are more like medium-impact themselves), and a corresponding spectrum of diligence and review expectations.

And yes, from time to time someone makes a wrong call and merges too a risky change too quickly and then there are reverts and general friction.

But also some details of the policy will impact specifically the long tail because in higher-impact areas things will need to pass more scrutiny regardless of the origin of the lines of code.

Author: I think these considerations should mostly balance each other out, right? The baseline expectations apply to every contribution, and since more important code (ideally) gets more scrutiny, cases where the baseline expectations are insufficient to ensure the required level of quality should be more likely to be caught.

I agree with @reckenrode that we shouldn’t try to explicitly delineate a core vs. peripheral distinction and apply different standards to each, in the context of a formal policy.

> It is not permitted to submit automated contributions without any manual review or intervention, outside of standard community automation.

Reviewer 27: I have thought about making automated PRs, as an alternative to (smaller) treewides where there is worry about functionality (e.g. removing dead code, where it's often not clear whether the code _should_ be dead or alive). The advantage would be that, with one PR per package (or per unique set of maintainers, or per, like, 5 packages), the maintainers could be requested for review, and request changes, approve, or even merge themselves with the merge bot. (And, because it would be mine, I would be able to easily look at PRs, respond to feedback, close, merge, etc.)

I haven't actually created this. But I think it would be useful in some circumstances, and so I'd be reluctant to prohibit it entirely?

Reviewer 8: off-topic (?): i think i've seen considerations (by @infinisil?) on using a single PR for tree-wide changes making for easier reverts - tho for all i know this might depend on the topic as well

Reviewer 15: Also, if a maintainer has an objection appplicable to multiple packages in the change, it is a bit annoying to paste it to all the relevant PRs… Unique set of maintainers is not really enough to guarantee low amount of per-maintainer spam. If a specific maintainer agrees, one can break out a specific subset of changes after their review comment.

Author: It sounds to me like this is a process you’d drive with yourself in the loop for a specific treewide, right, @mdaniels5757? I think that doesn’t count as the fully‐automated zero‐intervention no‐oversight bot case this is intended to rule out.

Reviewer 27: emilazy: Yes, that's correct. OK, good to hear.

> Automation without any manual review must not be used as the sole arbiter of whether to merge a change.

Reviewer 27: @numinit Would NixOS/rfcs#200 run afoul of this? :)

Reviewer 29: Amateur radio stations are required by law to be operated with a human in the loop, and the nixpkgs-review MCP only allows running reviews and replying with NAK and LGTM. So:

`.--. .-. --- -... .- -... .-.. -.-- --- -.-`

Reviewer 30: I believe that `nixpkgs review pr --post-result` counts as "standard community automation". The output is clear that it is an automated run, analogous to a manually invoked build bot. It should also, as usual, not be the sole arbiter of whether to merge.

(Yes, I saw the date on that. The point still stands.)

## Transparency

All covered use of automated tooling for a contribution must be disclosed as part of that contribution.

> In the case of LLM‐based AI tooling used for commits, this **must** be in the form of a Git commit trailer in the format `Assisted-by: AGENT_NAME:MODEL_VERSION`.

Reviewer 28: I feel like "agent name" is a little too vague. "Agent name" could be a harness that was used (opencode, Codex etc.), it could be some sort of nickname or codename the author uses for their *agent* (a combination of harness, language model or model family, system prompt and perhaps a memory system in case of autonomous persistent agents).

Additionally, `Assisted-by` could (and, in my belief, *should*, though for purposes of this policy I'd put in an RFC2119 "MAY") be used for other tools, not just language model-based coding assistance agents.

I seem to recognize that the format was copied from the Linux kernel coding assistant guidelines. I think we should just copy that section in full (with explicit acknowledgment and backlink):

> ```
> Assisted-by: AGENT_NAME:MODEL_VERSION [TOOL1] [TOOL2]
> ```
>
> Where:
> - `AGENT_NAME` is the name of the AI tool or framework
> - `MODEL_VERSION` is the specific model version used
> - `[TOOL1] [TOOL2]` are optional specialized analysis tools used (e.g., coccinelle, sparse, smatch, clang-tidy)
>
> Basic development tools (git, gcc, make, editors) should not be listed.

Reviewer 28: Also I think it would be nice to clarify behavior if several harnesses/models were used. One `Assisted-by` trailer for each model/harness pair?

Reviewer 31: Some examples of `AGENT_NAME:MODEL_VERSION` could be useful, as there is still ambiguity.

For example, what would be the best form amongst the following?
-  `opencode/unsloth/Qwen3.6-27B-GGUF:UD-Q4_K_XL`
- @codex`/Qwen3.6-27B`
- `github.com/openai/codex/Qwen3.6-27B`
- `github.com/openai/codex/unsloth/Qwen3.6-27B-GGUF:UD-Q4_K_XL` (and we are lost in the `/`).

Reviewer 28: > For example, what would be the best form amongst the following?

@hoh all of the examples you listed below have a `/` where a `:` should be per the policy described above, and the 1st and 4th example also use a colon for quantization specifiers (the only place where a colon is used in model releases anyway).

All of this definitely requires clarification, and official examples would be useful.

I imagine it like:

- `opencode:qwen3.6-35b-a3b` (downsides: no quantization specifier)
- `opencode:qwen3.6-35b-a3b:q4_k_m` (downsides: none, but watch out for the second colon if parsing via automation)
- `opencode:unsloth/qwen3.6-35b-a3b-gguf:q4_k_m` (downsides: a little more verbose, though this form can be useful if using obscure finetunes/quants/variants; implied huggingface link)
- `opencode:hf.co/unsloth/qwen3.6-35b-a3b-gguf:q4_k_m` (downsides: a lot more verbose, but extremely specific; I could paste this into llama.cpp and get the exact same model you're running, save for the sampler parameters, which seem to be a matter of taste anyway, it seems)
- `github.com/simonw/llm:<any of the above>` (downsides: **extremely long**, but also has a direct link to the harness. One could even do a specific commit, or point to a fork if harness is customized; still doesn't fully capture the workflow, and can't hope to, because some harnesses allow plugins for a total conversion into something else entirely; also, excludes non-OSS harnesses, which to me feels like a good thing, but some proprietary/unpublished harness users will rightfully object, and soon I may be among the latter)

Reviewer 31: When reflecting on this topic, I am wondering what the real value is for mentioning the specific tools and models used.

A user may also be using a collection of tools, such as:
- Cursor with Opus 4.7 for creating a plan
- OpenCode with Qwen 3.6-35b-a3b for implementing it
- Codex with GPT 5.5 for reviewing the changes
- GitHub Copilot for a second code review
- Claude code with Haiku for fixing typos

There are also web versions of Codex/Claude/...

This would end up with a very verbose list of advertising for the companies behind these tools, but I don't see the value above something like `Assisted-by: Coding agent`.

Author: I think that it’s good to establish provenance here as part of transparent disclosure, and I don’t think it makes sense elide them because of considering them marketing any more than it does to treat disclosure of paid contributions the same way. I don’t see a world in which we’d regret recording this data, and I can think of multiple ways we might end up glad to have it on the record; we should certainly be prepared for certain tools or models to potentially pose more problems than others, I think, and [GHSA-67f2-674w-6g63](https://github.com/NixOS/nixpkgs/security/advisories/GHSA-67f2-674w-6g63) has given me a strong appreciation for maximalism in audit trails. (It also matches what other projects are doing.)

The exact format probably does not matter all that much, but we should likely clarify it further, yeah.

Reviewer 16: > but we should likely clarify it further, yeah.

I think there's a good reason to leave it vague, with some examples. The landscape is extremely fresh at the moment. We went from nothing to models to versions to quants to effort levels to harnesses to skills to agent coordinators to frameworks in a very short time. Trying to fit that into a specific format now will likely make it confusing and outdated again in 6 mths.

Alternative option: list a few examples including a many elements and encourage being detailed, so for example "opencode + superhuman + gpt 5.5 xhigh + openspec". At least the "include more rather than less" will allow for natural expansion and the example will set the expectations.

Reviewer 8: seconding @hoh here - the prevalence of e.g. claude's suggested `opusplan` default (have the bigger model handle planning mode, the mid-sized one do implementation) already makes this format awkward.

this further raises the question of what exact problem we were trying to solve with that detail. it's not like model details make things reproducible, while i imagine the reason LLM vendors had the `Co-Authored-By` tags include any info on this was marketing, basically.
like, should we go along with that and make nix contributors do marketing for LLM vendors just because big AI wanted us to?

Reviewer 15: We might end up collecting a bunch of «not even the top models are reliable» examples of mess-ups, and maybe a few guidelines of type «if you use this, check this and document the check or get the PR auto-closed for carelessness». And yeah, maybe we might end up needing to emergency revert as much as possible of commits touched by a specific setup.

Reviewer 8: > We might end up collecting a bunch of «not even the top models are reliable» examples of mess-ups

tbh LLMs are 'garbage in garbage out' w.r.t. prompts as well as to prior context, and either way should be challenged by someone with some domain knowledge on suggested solutions, if not further to conversely have the machine challenge the user on ideas and implementation as well.

to an extent then, i would be inclined to regard any use of a tool settling for a shitty output as using it wrong, such that i would argue having some users publish shitty output should not have to imply another wouldn't be able to use it better.

Author: (Joining threads: I’ve responded to some of the points raised here in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3213309794.)

Reviewer 15: > to an extent then, i would be inclined to regard any use of a tool settling for a shitty output as using it wrong, such that i would argue having some users publish shitty output should not have to imply another wouldn't be able to use it better.

On the one hand, sure, «get 70% there and just edit the last part by hand» (possibly with another round of fully-hand-handled LLM review) is solid advice so one could say that anything that gives worse results than that is holding-it-wrong. On the other hand, some data on how bad things can or cannot get with an allegedely vibecoding-safe setup if one doesn't pay attention could still be useful.

> In the case of LLM‐based AI tooling used for commits, this **must** be in the form of a Git commit trailer in the format `Assisted-by: AGENT_NAME:MODEL_VERSION`.

Reviewer 32: I am quite opposed to  `Assisted-by:` and `Co-authored-by:` tags in commits, because it helps LLM companies. My suspicion is that coding-focused LLM tools started using this tag pattern to help themselves avoid the so-called Model Collapse problem, where models trained on the output of other models leads to a drop in model efficacy. By instituting a policy of requiring such a commit trailer, we are opting-in to _helping_ the very LLM companies that are besieging the internet with their relentless crawling, by giving a positive and policy-mandated confirmation of commits they shouldn't incorporate into their latest round of training.

EDIT: realized I committed review faux-pas of not including thoughts on remediation.

Beyond my personal ideal of auto-closing any PR which contains these commit trailers, the only thing that I can think of would be to instead require a similar string in PR description bodies. It'd still be straightforward to consume in an automated manner for Nixpkgs' management purposes, and offers at least one level of indirection to not so explicitly help LLM companies. Downside is that it may still (eventually) introduce that same consistent indicator in the Git commit history, but at least it's not _quite_ as regular.

---

Apologies if this has already been mentioned elsewhere. I searched through comments but wasn't able to find anything.

Reviewer 9: How about Assited-By: but without brand names

Reviewer 9: The concern is fair, and was brought up e.g. at https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3166094435: we're picky about who we promote at nixcon, why not in git log...

I like git-blames and would like to be able to roughly assess which lines come from contributors I know through how many handshakes. Similarly I'd like to know roughly which lines are generated. Metadata in commit messages mostly does the job. `Co-authored-by: klanka` as the baseline. (further discussion about what other metadata to include: https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3158099804)

Reviewer 15: An argument for PR description is also that this is more visible on review. I guess we could also say explicitly that package maintainers may establish stricter constraints on generated code in their area of responsibility than the bare minimum baseline requirements over Nixpkgs?

Additionally, maybe we should say that PR authors MAY add Assisted-by commit headers, for whatever they consider a tool. To make signal expensive to use at scale, one needs deviations from naive expectations in both directions, after all.

Reviewer 8: @SomeoneSerge:

> Co-authored-by: klanka

nit: while adding a layer of indirection, one could argue adding such metadata might still not fully address the concern raised by @WillemToorenburgh here, tho i'd concur on removing the advertising

Author: (Joining threads: I’ve responded to some of the points raised here in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3213309794.)

Reviewer 33: > I am quite opposed to `Assisted-by:` and `Co-authored-by:` tags in commits, because it helps LLM companies. My suspicion is that coding-focused LLM tools started using this tag pattern to help themselves avoid the so-called Model Collapse problem, where models trained on the output of other models leads to a drop in model efficacy. By instituting a policy of requiring such a commit trailer, we are opting-in to _helping_ the very LLM companies that are besieging the internet with their relentless crawling, by giving a positive and policy-mandated confirmation of commits they shouldn't incorporate into their latest round of training.

This is largely not true. The "model collapse" problem as it is commonly described is a myth. Current SOTA LLMs are actually trained on large amounts of synthetic data, including data generated by older models or previous versions of the same model. While it's not public how this data is used, there are examples of open weights models like Arcee Trinity that use as much as 50% synthetic data during pretraining. This is a big part of how LLMs have continued to improve even as novel human-created data has become harder to find. It's also why OpenAI and Anthropic are always complaining about Chinese labs distilling from their models (i.e. using ChatGPT and Claude outputs as training data for their own models). They wouldn't be upset about it if it didn't work!

So tagging AI-assisted commits is not helping AI companies, because they are likely not avoiding AI-generated code to begin with. Of course, if they wanted to avoid it, I'm sure they have more reliable AI detectors for internal use which do not rely on bespoke commit footers. I don't think it makes sense to not require marking AI-assisted commits just because of a small chance it would slightly benefit AI companies, when requiring them seems like it would almost certainly benefit nixpkgs maintainers much more.

Reviewer 32: If what you're saying is accurate, then I've been mis-informed, and would withdraw my concern from the discussion. Thank you for the new information!

A `Co-authored-by:` trailer does not satisfy this policy.

Any adequate form of disclosure is permitted for other kinds of tooling and contribution.

> Pull request summaries must be disclosed separately to commits.

Reviewer 8: In summary of the two discussion threads here, here is a version foregoing the (more machine-readable) commit-level provenance in favor of transparency aimed at humans:

```suggestion
All covered use of automated tooling for a contribution must be disclosed as part of that contribution.

Any adequate form of disclosure is permitted, such as disclaimers in the pull request.

`Co-authored-by:` commit trailers are discouraged over their use to market specific tools.
```

Reviewer 32: Well done. I would find this much easier to swallow, should it end up in the final copy

Reviewer 9: I'm in favour of the document suggesting an explicit marker (`Assisted-by: whatever`) that makes it easier to automatically filter the commits, but discouraging ads

Reviewer 8: @SomeoneSerge could you submit a formal code edit suggestion here to that extent, so we could gauge relative support?

Author: I continue to believe that disclosure is the foundation of any effective policy. I made many of my points in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175196784, but I want to expand on them since this has come up a few times. (I’ll address points from both threads, https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3158099804 and https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3197975810, to try and keep the discussion on this topic centralized.)

I really want to challenge the idea that the purpose of transparency is promotion. Putting corporate logos front and centre as valued sponsors on a community‐facing page with money exchanging hands, as mentioned in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3198153352, is a very different matter to transparency requirements as part of our contribution workflow. We have no restrictions on the disclosure of funding for contributions, even from companies that have been declined for sponsorships, and I think we should encourage it more so we can increase transparency and take conflicts of interest into account. Some companies require employees to use corporate emails or include disclaimers to contribute to FOSS to signify their claimed copyright interest in the work; those are valuable for licence compliance matters and shouldn’t be banned on the grounds of advertising. We have informal precedent for expecting people to disclose the automation they’ve used for certain kinds of changes, but no precedent for preventing or discouraging people from explaining what tools they used to create a change.

Separate disclosure of commits, pull requests, and comments is important to uphold the communication norms discussed in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175207606, and we should try to avoid have load‐bearing attributions for the provenance of source code locked up solely in GitHub and unavailable when viewing commits in the repository. But even an `Assisted-by:` trailer without any detail would, I think, be a mistake. We have seen this tooling change drastically over a short period, as per https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3176501203, and a foundational principle of this policy is that the degree and nature of the automation being used for a contribution has a meaningful effect on how reviewers will want to approach it.

It’s also relevant to future policy work; if we are going to review how this initial policy goes and amend it in response to the data we collect as a result of the transparency requirements, knowing what kinds of tools are being used and how, and being able to make apples‐to‐apples comparisons about how it’s going, will be vital. I also agree with https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3182250399 that such an audit trail is important not just for our own visibility but for potential future legal compliance.

I do agree with the points raised that the specific format currently listed in the policy is currently an awkward balance of overly constrained and underspecified, and we are working on amending it.

Of course, any discussion of any company or product promotes awareness of it to some extent, but we don’t set our communication standards based on that; honestly, I’d say Nixpkgs packaging corporate software advertises far more than attribution for a functional purpose as part of development does. I don’t think that custom commit trailers in Nixpkgs that don’t surface in the GitHub UI is a particularly valuable advertising space for companies at the scale we’re talking about here, and even omitting specific tool and model names wouldn’t eliminate concerns about advertising; insofar as disclosures do act as advertising, even a bare minimum disclosure of “LLMs were used” required to serve the goals of the policy would serve as advertising for the industry as a whole, and with few major players in the space would likely be comparatively effective. I believe we shouldn’t set policy that puts us in a worse position just because there is always some unavoidable element of promotion in any kind of transparent disclosure. Even having such extensive policy discussions about this serves as a kind of promotion, after all, and yet is clearly worth it.

I don’t think I could stand behind a policy that weakens the transparency requirements to the extent being suggested in these threads.

Reviewer 9: > ...knowing what kinds of tools are being used and how, and being able to make apples‐to‐apples comparisons about how it’s going, will be vital

Like, yes, but this ship has sailed when the industry has given up on training their own models and settled for either renting the Cloud options, or running distilled models. It would be completely useless to type `Assisted-by: opus4.7`, because `opus4.7` is not one fixed thing, but some hundreds of versions of the model and the harness redeployed under the same release name and behind the same API. I would _love_ to have each generated commit to contain an `Assisted-by:` with a (non)flake-ref to the model, describing the data, the hardware, the schedule that was used to produce it, but there's literally nobody who does that today afaik

Reviewer 8: on the higher levels of transparency alluded to by @SomeoneSerge, NLNet [has a policy](https://nlnet.nl/foundation/policies/generativeAI/) mandating commit descriptions include full interaction logs. while i'm not sure that was necessarily written with coding agents in mind, it might further increase commit size as well.

Reviewer 32: Truthfully, the advertisement component never entered my mind when I was writing my initial comment. For me, it's all about not making Nixpkgs overtly helpful to LLM companies. On the other side of the coin, I _deeply_ understand the desire to avoid vendor lock-in with LLM usage disclosures only living in GitHub pull request bodies. I appreciate that this is a discussion, and the effort that the maintainers are putting into it.

It's clumsy and more human, but what if we had something like an `AUDIT/` directory, where each PR has a corresponding `AUDIT/GH-<PR number>.md` file containing at least the title and body of the PR (perhaps prompt transcripts too), and a requirement for the file to be included in a given PR when an LLM was used? Such a requirement could possibly be made easier to comply with via some scripts... but this is almost certainly scope creep.

Reviewer 15: > either renting the Cloud options, or running distilled models.

I would say that in terms of specifying the version, distilled models run locally are at least pinnable, and often base / distillation / fine-tuning combination is roughly known, and we'll see if provenance based on weight signatures ends up working. If it is not local, we just have a case of Cloud with all of its usual problems and misleading naming.

## Exemptions

The following situations are fully or partially exempt:

> * Use of standard deterministic editor/IDE/formatter/text transformation tooling to produce changes that the author manually reviews and understands is exempt, including inline “auto‐completion” (even if LLM‐based) of short, rote snippets of text that do not contribute anything beyond boilerplate the author would have written anyway.

Reviewer 10: ```suggestion
* Use of standard deterministic editor/IDE/formatter/text transformation tooling to produce changes that the author manually reviews and understands is exempt, including inline “auto‐completion” (excluding LLM‐based) of short, rote snippets of text that do not contribute anything beyond boilerplate the author would have written anyway.
```
I actually think its far too easy to have inline LLM autocomplete to write code for people who don't actually understand what is happening. Even if they do, I think it is still worthwhile to make it known in the trailers

Reviewer 9: :+1: I think it's this kind of interactive "auto-completion" that it's not immediately obvious we need a moratorium on, but that is susceptible to "getting asleep at the wheel" and where we need contributors to pay special care to "establishing confidence"

Reviewer 24: LLMs _are_ autocomplete - that's literally what they are designed to do. Making an exception for autocompletion is effectively exempting all LLM output.

Reviewer 9: https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3167058931 for 1 more spin at this intuition: what we want to require is the "slow mode", i.e. that the reviewer is not exposed to the "generator"'s artifacts at a pace any faster than their normal writing and review speed

> LLMs _are_ autocomplete

Yea I think we're referring to the "LLMs"-the wrappers around "LLMs"-the autocomplete blocks.

Reviewer 20: The text already limits this to boilerplate.
If that's unclear, perhaps include a counterpoint?

```suggestion
* Use of standard deterministic editor/IDE/formatter/text transformation tooling to produce changes that the author manually reviews and understands is exempt, including inline “auto‐completion” (even if LLM‐based) of short, rote snippets of text that do not contribute anything beyond boilerplate the author would have written anyway. (So it follows that that generating any actual implementation logic is *not* exempt, regardless of how the automation is interacted with.)
```

Author: At least JetBrains IDEs automatically show short LLM‐based autocompletions out of the box; I expect it probably doesn’t apply to Nix code specifically, but in general I would suspect that we’ll see features like that deployed more widely, and it’s not clear to me that the degree of harm posed by them counterbalances the downsides of having a policy that catches a bunch of very benign, entirely manually‐directed changes, and that people might reasonably not realize they’re out of compliance with even after reading the policy.

I agree that it doesn’t rule out complacency at all, but when strictly scoped to short snippets of rote boilerplate, I’m not sure it’s a huge qualitative change compared to pre‐existing types of tooling, in the way that automated coding agents are. (I saw [TabNine](https://web.archive.org/web/20181106181527/https://tabnine.com/) talked about a little when it came out many years ago, but not huge amounts of hype or a sea change in the observed dynamics of FOSS contributions.)

The restriction is also meant to be strict enough to rule out e.g. any worry of copyright concerns. (To a large degree it is likely redundant to the “generate *non‐trivial* amounts of output” wording – or elaborating upon it, depending on how you look at it.)

I agree with @roberth that “short, rote snippets of text that do not contribute anything beyond boilerplate the author would have written anyway” is the load‐bearing criterion here, not “auto‐completion”, and certainly doesn’t exempt all LLM output. But if there’s a way we could clarify the former further to make the tight scope clear, that would be good.

> * Use of standard community automation is exempt, such as `nix-update`, the official Nixpkgs CI bots, the @r-ryantm update bot, and the Nixpkgs security tracker bot.

Reviewer 17: How does this affect custom update bots as some maintainers have used them in the past ([including me](https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+author%3Abotnk+is%3Aclosed))? Are these still allowed to create PRs on their own without the operator's review before undrafting, or are they explicitly addressed by this policy (i.e, operators are meant to review the automated PR before it is undrafted)?

Context for my use of this (currently retired) bot: zed-editor has a rather short release interval, and for a while there were no other contributors willing to take care of update PRs (which is not the case anymore), so I tried to automate the process in the meanwhile. Also r-ryantm does not notify me of issues in the update process, which my own bot did. I personally think requiring such automated PRs to be drafts until the operator has verified their output is reasonable though.

Reviewer 20: Perhaps include `updateScript` and tooling that's created or approved by the maintainers of the individual package?
We can spread the decision making here.

```suggestion
* Use of standard community automation is exempt, such as `nix-update`, the official Nixpkgs CI bots, the @r-ryantm update bot, and the Nixpkgs security tracker bot.

* Use of automation approved by the maintainers of the affected set of files, provided that any automated communication such as PR descriptions and automated commit messages are clearly identifiable as the result of that particular automation as laid out in [Transparency](#Transparency).
```

(without enforcing that agreement to be formal, so I think it should be fine for a solo maintainer could just go ahead)

Author: Custom update bots seem like a reasonable use case, and the fact that they’re serving a specific maintainer who has opted in to seeing the PRs certainly helps. However, it seems tricky to delineate the boundaries here. I am not sure we would want to give the green light to, say, LLM‐based bots submitting automated changes en masse, even with maintainer consent, even as drafts, which it seems like @roberth’s amendment would do.

I think that “standard community automation” ought to include update scripts, but update scripts themselves don’t open PRs, which runs into “It is not permitted to submit automated contributions without any manual review or intervention, outside of standard community automation”. Maybe we could have a specific carve‐out here by simply generalizing “the @r-ryantm update bot” here to e.g. “bots that run in‐tree update scripts”?

It’s of course also an option to have these bots present these changes to the relevant maintainers outside of Nixpkgs (say in a fork), and then to migrate them to Nixpkgs once the maintainer has approved, which would meet the policy requirements. But I realize that complicates the workflow, so it would be good to find a way to permit this as‐is.

* Use of AI tools for research, testing, debugging, or private review is out of scope, if no substantial amount of their output is included in the resulting contribution.
  However, if these tools had a significant technical influence on your contribution, you are still responsible for it per the Accountability section, and are expected to disclose this where relevant.

> * Use of machine translation is exempt from the requirement to understand the translated output.
>   However, the requirements of appropriate confidence in the original text, responsibility, and disclosure still apply, and you are encouraged to additionally include the original untranslated contribution.

Reviewer 9: I believe this isn't even an "exemption" as such, more of a clarification of how we interpret what machine translation does. Rather than reasoning about "interactiveness" we could talk how direct or how "invertible" the correspondence between the inputs and the outputs is.

```suggestion
* Machine translation is assumed to be sufficiently interactive for the contributor to retain their understanding and confidence in the generated outputs.
  The accountability and disclosure policies still apply, and you are encouraged to include the original untranslated contribution.
```

Reviewer 8: i wouldn't personally use interactive as the distinguishing factor, in the sense the process of use of coding agents one might argue to include interactive elements as well

Reviewer 9: Fair point, the interactivity I'm talking about is "you are forced to come up with and spit out every next hunk at exactly the same pace as the model"

Reviewer 20: Machine translation *of human language*, right?

"Appropriate confidence" is load bearing here, but good enough IMO.
I'm inclined to go on a tangent to say that i18n is hard, but that's (unfortunately?) mostly irrelevant for Nixpkgs anyway for now.

Translation is not where the difficulty is, so the current text looks sufficiently vague to me so as not to pose a problem.

Reviewer 9: > Machine translation of human language, right?
> ... i18n ... hard

Si, and si. "Slow mode" is my current proxy for what we're trying to pinpoint here: https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3167076386

Author: I think this wording is likely to be more confusing for readers, as it covers the justification for providing an exemption while leaving the actual exemption implicit. The review requirements are explicitly about the output; most cases where automation is processing a contribution into a form that is no longer comprehensible to the contributor *would* be a problem, I think, and machine translation has enough pitfalls that I wouldn’t say we can make a general assumption that the contributor understands its output – hence also the suggestion to include the original text.

Reviewer 9: > automation is processing a contribution into a form that is no longer comprehensible to the contributor would be a problem, I think, and machine translation has enough pitfalls that I wouldn’t say we can make a general assumption that the contributor understands its output

Precisely. That's why I'm saying it's not really an exemption, nor does it need to be one. The paragraph is just to explicitly state, that the use of translation and accessibility tools in itself is not automatically disqualifying.

  This does not waive any other requirement.

* Use of automated tools to develop upstream software packaged inside Nixpkgs is not in scope.

## Enforcement

If you believe that someone is using automation without appropriate disclosure and review, you can politely ask them if that’s the case and point them to this policy as appropriate.
Please assume good faith and remain civil; it’s not always possible to determine, and it is more likely that someone overlooked this policy than deliberately violated it.
If you think someone is continuing to break the policy after this, please escalate to the [Nixpkgs core team](https://nixos.org/community/teams/nixpkgs-core/) rather than fighting over it.

Deliberate violations of this policy are considered to break the [Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) clause against “Wasting other people’s time with low quality contributions, including but not limited to LLM and bot spam”.

> If a contribution clearly violates the policy – e.g. the contributor admits it is in violation without working to fix it, or there are AI tool attributions that do not meet our required format – it is acceptable for anyone with technical permissions to close or hide it, pointing to this policy if necessary.

Reviewer 17: To clarify, is closing a PR in case of "AI tool attributions that do not meet our required format" exempt from the above "assume good faith"? More specifically, are such PRs that use `Co-Authored-By` for AI disclosure always to be closed directly without giving a chance of correction?

Reviewer 6: I hope not, especially for first-time contributors who otherwise seem to be acting in good faith -- granted, I might be exactly the kind of contributor that you want to drive away with this as someone who has admittedly been a bit too liberal with coding agents in some of my PRs, and in that case I'll take my lumps and step away from the discussion, but I really do think that continuing to be a welcoming place for new contributors is worthwhile, even in a world where those new contributors will increasingly make at least some use of AI-based tools and a lot of people in the community have strong feelings about that.

Reviewer 4: To me this paragraph reads like the idea is to grant people who inadvertently violated this policy an opportunity to fix their mistake.

Reviewer 20: I worry a bit that this paragraph could be used to justify unwarranted behavior on the part of maintainers. I'd favor a phrasing that keeps the temperature low, and we could adjust the policy if we end up becoming more of a target for slop somehow, and even then, the CoC already covers this, albeit with fewer specifics.

Furthermore if this scares away LLM users in the first place, that would be bad.
- Below the generated code are usually real issues that are worth solving
- LLM users contribute real-world testing
- Improving open source together should reduce LLM usage and its negative externalities (i.e. avoid duplicated LLM work between many users)

I'd consider dropping this paragraph and only reintroducing it when proven necessary.

Author: The assumption of good faith is not meant to be optional here, of course!

By the definition of [extractive contributions](https://llvm.org/docs/AIToolPolicy.html#extractive-contributions), however, we need a way to handle them that requires sufficiently low friction on the part of the person doing triage; see [LLVM](https://llvm.org/docs/AIToolPolicy.html#handling-violations) and [Rust](https://github.com/rust-lang/compiler-team/issues/893) for concrete precedent on a “quick rejection” flow for PRs that don’t meet standards set by related policies. I believe that we cannot omit some clear indication that contributions that don’t meet the requirements aren’t accepted, and that it’s fine to deal with them quickly and move on rather than debating the policy every time.

We already have a significant triage problem; a dramatically lowered cost of submitting plausible‐looking contributions exacerbates it, which is one of the primary motivations for this policy. In general, I think Nixpkgs is quite bad at saying “no” to things; contributions tend to stall out rather than being closed, which I think is no less frustrating for contributors.

I agree, though, that this shouldn’t come at the expense of being polite, or giving the contributor a chance to address the problem. The ideal flow I imagine looks something like this:

1. We have a label for contributions that don’t meet the policy. CI detects contributions that can be unambiguously determined to violate the policy (e.g. `Co-authored-by:` without `Assisted-by:`) and applies it; contributors can also manually apply it where relevant.

2. CI detects the label, marks the PR as a draft, and posts a comment explaining the need to meet the policy and pointing the contributor to it.

3. When the contributor has adjusted the contribution to meet the policy, they can check the PR checklist item, and the PR will be undrafted and the label removed.

4. If the contribution isn’t adjusted within a week, CI closes it automatically.

I think this should allow us to minimize the burden posed by contributions that are in violation of the policy, while being kind to contributors. We’ll take a look at the specific wording here, though, as we don’t want to encourage reviewers to cause conflict, especially unilaterally in ambiguous cases.

Repeated violations are grounds for further moderation action.

## Credits

> This policy takes inspiration from similar policies in [LLVM](https://llvm.org/docs/AIToolPolicy.html), [Mesa](https://gitlab.freedesktop.org/mesa/mesa/-/blob/mesa-26.1.0-rc1/docs/submittingpatches.rst?ref_type=tags), [Fedora](https://docs.fedoraproject.org/en-US/council/policy/ai-contribution-policy/), and the [Linux kernel](https://docs.kernel.org/7.0/process/coding-assistants.html), along with [a proposal by the author of Anubis](https://xeiaso.net/notes/2025/assisted-by-footer/).

Reviewer 9: Here's a good example of "playing safe" (thank you `@xhalo32` for the link):

> Note that using an LLM when writing comments on github or Zulip is not allowed: use your own words.
> https://leanprover-community.github.io/contribute/#use-of-ai

I believe Mathlib's decision is particularly relevant to us, because it's a very similar project: it's a holistic and orderly compilation of our collective human knowledge about everything-mathematics in a single unified language, same as Nixpkgs is a compilation of collective human knowledge about everything-software. Projects like Mathlib and Nixpkgs (...Wikipedia) are precisely where we need to establish a "blood-brain barrier" and prevent parasitic feedback loops: "Nixpkgs is an [input, not an output](https://discourse.nixos.org/t/policy-on-vibe-coded-nix-tools/76300/26)"

Reviewer 15: Mathlib is about maths, an subject with pretty deep respect for orderliness. Much of Nixpkgs is about software build systems when nobody is intervening in time to prevent horrors, a chaotic subject in practice. The closer to core you get, the more orderly people try to keep things, but at the periphery, sometimes «one who fights monsters becomes a monster». 

> prevent parasitic feedback loops

Our real needs are about working software, if chatbots stop working, no chatbot contribution will pass the basic requirements, consensus finally achieved. And whether chatbots do stop working should not be our problem to fix. People who want to make chatbots work will do, people who want to poison chatbots without damaging the actual functionality of Nixpkgs will do, that's fine.

Reviewer 9: > Much of Nixpkgs is about software build systems [...] a chaotic subject in practice

Yes, Nixpkgs is "applied maths".

> Our real needs are about working software,

..."working software" here meaning "working in the long(er) term", i.e. "bootstrapable, maintainable, inspectable", of course.

> People who want to make chatbots work will do, people who want to poison chatbots without damaging the actual functionality of Nixpkgs will do, that's fine.

Chatbots are supplied with enough money to keep chugging forward and getting subscriptions regardless of quality of their underlying models. It is Nixpkgs as the social phenomenon and as a world-modeling endeavor, with its Wikipedia-style development model, that is at risk.

Whether we're talking about Nixpkgs-for-deploying-today, or Nixpkgs-for-reading-by-humans, or the Nixpkgs-for-data-analysis, all three are of most value when written by humans.

  However, if these tools had a significant technical influence on your contribution, you are still responsible for it per the Accountability section, and are expected to disclose this where relevant.

* Use of machine translation is exempt from the requirement to understand the translated output.
  However, the requirements of appropriate confidence in the original text, responsibility, and disclosure still apply, and you are encouraged to additionally include the original untranslated contribution.

* Use of automation in a contribution clearly marked as not being ready for merge (e.g. a draft pull request) is exempt from the requirement for full self‐review, as long as some amount of review has been done and it is expected that the requirements will be met by the time it is marked as ready.
  This does not waive any other requirement.

* Use of automated tools to develop upstream software packaged inside Nixpkgs is not in scope.

## Enforcement

If you believe that someone is using automation without appropriate disclosure and review, you can politely ask them if that’s the case and point them to this policy as appropriate.

> Please assume good faith and remain civil; it’s not always possible to determine, and it is more likely that someone overlooked this policy than deliberately violated it.

Reviewer 34: I'm concerned that this has the risk of being violated almost as much, and in concert with, the contributions of LLM tools.

Reviewer 28: @adamcstephens I think at least explicitly putting up a standard for LLM contribution disclosure may still help to shrink down the amount of witch-hunting.

In my opinion, this policy is better than an outright ban (which would *encourage* witch-hunts and treating every new contributor with suspicion) and better than blanket ignorance (which may lead to the anti-LLM part of the community taking the matters into their own hands in an attempt to sabotage LLM tooling; this, unfortunately, already got a precedent despite the policy being in the works, see #516544).

The policy as-is now states that LLM contributions are welcome, if they're of sufficiently high quality and appropriately disclosed. Which in effect can be treated as:
 - non-disclosed "slop" PRs can be just closed on "bad quality" and/or "non-disclosure of LLMs" if the latter is glaringly obvious
    - Obvious incorrect disclosure would be the "Co-Authored-by:" advertisement trailers that harnesses love to inject via system prompt instructions
    - Otherwise I'd expect to close due to "bad quality", not "LLM witch-hunting". Judge by the content, not by the first impression.
 - non-slop appropriately disclosed PRs will serve as success stories for proper LLM usage and may serve as examples for others eager to adopt LLM-based automation in a way that doesn't reduce their contribution quality
 - appropriately-disclosed PRs of bad quality may allow identifying weaknesses in specific LLMs or setups, or lack of skill on part of the contributor. These are all fixable, if the contributor is willing to take responsibility and learn from their and their tool's mistakes.

Reviewer 34: Sorry if I was unclear. I agree this is all a good improvement. I support that we want to avoid bad contributions of all types, and LLMs make bad contributions easier. 

I guess I’m saying that assuming good faith and positive intent will require its own diligence, and possibly should be more prominent.

Author: I agree that there’s a risk of enforcement getting uncivil here and that would be a very unfortunate failure mode. I hope that a clearly‐defined policy and procedures for how to deal with violations will help reduce the temperature here, especially if we can get a CI‐based flow like I described in https://github.com/NixOS/nixpkgs/pull/514587#discussion_r3175200812 to handle things smoothly without taking too much of people’s energy. I’m also hoping that with the community team bootstrap group in place it’ll be more feasible to keep on top of these dynamics.

We amended the wording in this section already in response to the discussion at https://github.com/NixOS/nixpkgs/pull/514587#pullrequestreview-4196494710; do you have a suggested diff you think might improve things here? I’d be happy to tweak it further to emphasize this, but not sure if I have any good ideas other than reiterating “please do make sure to follow the CoC!” :)

Reviewer 34: No, I don't have a specific change to suggest right now. We'll see how it goes in practice and can adjust from there. Sorry for the noise.
